modules/ROOT/edits/dataweave-type-system.adoc

```diff
@@ -0,0 +1,357 @@
+----
+
+For example, if you want to define in a constraint a function type that has a parameter of type String, a second parameter of type Number and return a Boolean, the correct syntax is:
+
+----
+(paramA: String, paramB: Number) -> Boolean
+----
+
+For example, you could define a function that receives another function as an argument:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+fun applyIDsChange(ids: Array<Number>, changeTo: (Number) -> Number): Array<Number> = ???
+----
+
+If we call the function `applyIDsChange` with some function that does not match the `changeTo` constraint `(Number) -> Number`, DataWeave throws a type checking error.
+For example, the following call works because `abs` is a function that takes a Number and returns a Number, which matches the `changeTo` parameter.
+
+----
+applyIDsChange([1,-6, 3, -8], abs)
+----
+
+But this call fails because `sum` is a function that takes an Array and returns a Number, so it does not match our parameter constraint.
+
+----
+applyIDsChange([1,-6, 3, -8], sum)
+----
+
+== Complex Types
+
+=== Any and Nothing
+
+In some cases you can not enforce a restriction as it accepts all the values. +
+
+* `Any` type accepts all possible values.
+* `Nothing` type accepts no value but it can be assigned to all the types. This type is not frequently used explicitly but it is by the type inference algorithm.
+
+=== Union Type
+
+Union type is used to compose types. The syntax to define a union type is:
+
+----
+TypeA | TypeB | ...
+----
+
+The following example defines a variable with a constraint that accepts a value of type String or Number:
+
+----
+var age: String | Number = if (payload.allStrings) "32" else 32
+----
+
+A common pattern is to use Null with a union type to specify that a type accepts null as a value. In the following example, the variable `a` accepts that the expression `payload.foo` returns `null`.
+
+----
+var a: String| Null = payload.foo
+----
+
+
+=== Intersection Type
+
+The intersection type intersects object types. In this case, the intersection works as the concatenation (`++`) of object types.
+
+_Introduced in DataWeave 2.3.0. Supported by Mule 4.3 and later._
+
+The syntax for intersection type is:
+----
+TypeA & TypeB & ...
+----
+
+In the following example, the intersection concatenates the two object types resulting in the type `{name: String, lastName: String}` This is an open object so it can have more key value pairs. The variable accepts the value assigned to it:
+
+----
+var a: {name: String} & {lastName: String} = {name: "John", lastName: "Smith", age: 34}
+----
+
+In the case of close objects, it returns the concatenation of the object types but resulting in a close object. +
+In the following example, the intersection results in the type `{|name: String, lastName: String|}`. This throws
+an exception because a close object does not accept an object value if there are additional fields in the object (field `age`):
```

Comment:
```suggestion
an exception because a closed object does not accept an `Object` value if there are additional fields in the object (field `age`):
```